### ISSUE TITLE
Comparing deduplication/clipping to footprint with/without parallelization

### ISSUE THREAD

**Issue by julietcohen** [PermafrostDiscoveryGateway/viz-staging#19]
The deduplication / clipping to footprint seems to be working on the `viz-staging` branch `bug-17-clippingToFP` no matter what order the files are staged in (using 3 adjacent shp files as input). Keep in mind that this was not using parallelization. I would like to test if this branch deduplicates / clips to footprint with parallelization with ray on the Delta server, too, before I jump into making any other major changes to the branch.  

### Run on datateam

- not in parallel
- see notebook [here](https://github.com/PermafrostDiscoveryGateway/viz-staging/blob/1fd91634045e3e863e1c637c5dfae847ba15e7ed/pre-fix_3AdjFile_run/workflow.ipynb)
- produce 4376 geotiffs and 4376 web tiles for all z-levels

Web tiles visualized on local cesium show successful deduplication / clipping to footprint:
<img width="1154" alt="image" src="https://user-images.githubusercontent.com/88209596/225722198-52cb67a8-e65b-40a5-b5b4-8dd0caa88385.png">

Very small strips of no polygons between tiles indicates that clipping is working, but we should change the predicate from "within" to something else so we retain the polygons that overlap the footprint boundary rather than remove those as well as the footprints that fall completely outside the footprint boundary:
<img width="1326" alt="strips" src="https://user-images.githubusercontent.com/88209596/225722958-03f21fea-40e7-4bf6-9d92-04ef5c38f9d8.png">

### Run on Delta

Major errors with initializing ray. Couldn't get the `IN_PROGRESS_VIZ_WORKFLOW.py` script to start, hung at the `ray.init()` line.

**Comment by julietcohen** [comment]
### Run on delta with same 3 tiles

Got the run on Delta to work. Seems that the `ray_init()` issue I experienced last week was random (on Delta's end) because I did not change the Slurm script or way I initialized ray and it worked as expected! Wonderful. 

- run is to compare web tile output with web tile output from datateam run posted above that did not use parallelization
- Same 3 tiles visualized. 
- Used same branch of `viz-workflow`: `bug-17-clippingToFP`

Web tiles visualized:

<img width="943" alt="image" src="https://user-images.githubusercontent.com/88209596/226423573-bc9f5266-9ee9-4010-90b5-ec70a494cadd.png">

In a way, it's great that the tiles _were not deduplicated/clipped correctly_ with the ray workflow but _were with the non-parallelized run_ on datateam, because it helps us narrow down the source of the trigger for deduplication!

Differences between the runs were:
1. using ray versus not using ray
2. `viz-raster` branch: the datateam run without parallelization, where the tiles looked deduplicated/clipped, used the `main` branch of `viz-raster`, while the delta run with ray used the `bug-ID-notRecognized` branch of `viz-raster`, which was created specifically for the ray workflow to get around an error that was occurring where the ID object that is created by `__start_tracking` and used by `__end_tracking` could not be read. I think it is unlikely this is the cause of the trigger, because all I changed for this branch (see commit [here](https://github.com/PermafrostDiscoveryGateway/viz-raster/commit/4f1dfdc20dce4a1b9894c72c1ebb5148346d39fc)) was replacing that ID object with another string so we could still execute `__end_tracking`

**Comment by julietcohen** [comment]
### Staging step successfully identifies duplicate polygons both with and without parallelization

Ran script that opens all files in the `staged` directory, checks for any True values in the column we use to identify duplicate polygons in the staging step: `staging_duplicated`, and sums the number of True values. It was run on both the Datateam `staged` dir as well as the Delta `staged` dir that processed the same 3 adjacent & overlapping input files. (See comments above)

<details>
  <summary>check_if_any_true_dups.py</summary>

```
# Check if any staged files have True for the staging_duplicate property
# comparing delta staged data to datateam staged data 
# delta web tiles looked like they were not dedup
# datateam web tiles looked like they were dedup

from pathlib import Path
import geopandas as gpd
import glob

# ----------------------------------------------------------------------

# RUN THIS SCRIPT ONCE FOR EACH DATASET, COMMENT OUT ONE AT A TIME   

# DELTA DATA
# check if the staged files have any True values in their 'staging_duplicated' column
#input_dir = Path('/home/jcohen/compare_staged_dt_delta/delta_data/staged/')
#gpkg_files_posix = sorted(input_dir.rglob('*.gpkg'))

# DATATEAM DATA
input_dir = Path('/home/jcohen/viz-staging/pre-fix_3AdjFile_run/staged/')
gpkg_files_posix = sorted(input_dir.rglob('*.gpkg'))

# ----------------------------------------------------------------------

print(f"Found {len(gpkg_files_posix)} Delta staged files. Should be 3214 for both Delta and Datateam datasets.")

# remove prefix "PosixPath()" part from each full path
gpkg_files = []
for file in gpkg_files_posix:
    # extract just str filepath
    file_str = str(file)
    gpkg_files.append(file_str)

print("Removed Posix() part of all filepaths.\nChecking for True values in `staged_duplicated` col.")

true_values = []
false_values = []
# sum of len of true_props and len of false_props should be 3214.
# because that is the total number of staged files
# and the any() function returns 1 output per file
for file in gpkg_files:
    gdf = gpd.read_file(file)
    #values.append(str(gdf['staging_duplicated']))
    if any(gdf['staging_duplicated']):
        # if a True value is present in any row of this df's 'staging_duplicated' column
        # then add an observation to the True list
        true_values.append("True value detected")
    else:
        # if False is present every row of this df's 'staging_duplicated' column
        # then add an observation to the False list
        false_values.append("all False values detected")

print(f"Is the sum of the two lists ? {(len(true_values) + len(false_values)) == len(gpkg_files)}")
print(f"The number of staged files with at least 1 True value detected is {len(true_values)}.\n0 files means that 0/{len(gpkg_files)} polygons were identified as duplicates.")
```
</details>

In both cases, the number of True values (duplicate polygons) was 469. This is great, because running the workflow without parallelization and with parallelization identifies the same number of duplicates.

**Comment by julietcohen** [comment]
### Log file for non-parallelized run on Datateam

Notes:
- before the run, I inserted a logging statement to report when clipping is occurring
- also inserted logging statement within `clip_gdf()` to report number of rows it identified as `outside` and `within` the footprint (see [here](https://github.com/PermafrostDiscoveryGateway/viz-staging/blob/1fd91634045e3e863e1c637c5dfae847ba15e7ed/pdgstaging/Deduplicator.py#L88-L89))

statement | # of occurrences
-------------- | ------
"Clipping to footprint is being executed." | 938
"outside is: Empty GeoDataFrame" | 728
"within is: Empty GeoDataFrame" | 41

**This means:**
- clipping is being executed 938 times in total. Recall that the number of shapefiles fed into this run is 3, and the total number of resulting staged files is 3214. 
- some polygons were identified as falling outside the footprint 200/938 times that clipping to footprint was executed, and 728 times there were 0 polygons identified as falling outside the footprint
- some polygons were identified as falling within the footprint 897/938 times that clipping to footprint was executed, and 41 times there were 0 polygons identified as falling within the footprint
- This last statement is the most confusing. How could there be 41 tiles in which 0 polygons fall within the footprint?

**Comment by robyngit** [comment]
> This last statement is the most confusing. How could there be 41 tiles in which 0 polygons fall within the footprint?

It _might_ not be surprising that some tiles might contain just a small portion of a shapefile, and that portion only comprises areas outside the footprint.

If I understand your `check_if_any_true_dups.py` code correctly, it looks like it counts whether there are any true values for each tile. The next step might be comparing _how many_ true values there are for all tiles. Something like...

```python
import geopandas as gpd
import pandas as pd
from pdgstaging import TilePathManager

# The name of the column in the staged files that indicates whether a row is a duplicate
dup_col_name = "staging_duplicated"

# names for the two directories to compare (just for convenience)
dir1 = "delta"
dir1_path = "/home/jcohen/compare_staged_dt_delta/delta_data/staged/"
dir2 = "datateam"
dir2_path = "/home/jcohen/viz-staging/pre-fix_3AdjFile_run/staged/"

# Create a TilePathManager object in order to easily get the tile indices
# and staged file paths for a given tile.
tpm = TilePathManager(
    tms_id="WorldCRS84Quad",
    base_dirs={
        dir1: {"path": dir1_path, "ext": ".gpkg"},
        dir2: {"path": dir2_path, "ext": ".gpkg"},
    },
)

def get_tile_info(path):
    tile = tpm.tile_from_path(path)
    gdf = gpd.read_file(path)
    dup_col = gdf['staging_duplicated']
    return {
        "x": tile.x,
        "y": tile.y,
        "z": tile.z,
        "num_dups": dup_col.sum(),
        "num_rows": len(dup_col),
    }

files1 = tpm.get_filenames_from_dir(dir1)
files2 = tpm.get_filenames_from_dir(dir2)
df1 = pd.DataFrame([get_tile_info(path) for path in files1])
df2 = pd.DataFrame([get_tile_info(path) for path in files2])

df = pd.merge(df1, df2, on=["x", "y", "z"], suffixes=[f"_{dir1}", f"_{dir2}"])

df["num_dups_diff"] = df[f"num_dups_{dir1}"] - df[f"num_dups_{dir2}"]
df["num_rows_diff"] = df[f"num_rows_{dir1}"] - df[f"num_rows_{dir2}"]

print(f'Number of tiles with different number of duplicates: {len(df[df["num_dups_diff"] != 0])}')
print(f'Number of tiles with different number of rows: {len(df[df["num_rows_diff"] != 0])}')
```

If this shows that there are no differences in the number of duplicates identified, maybe the next step would be to look for differences in the max-z GeoTiff tiles.

**Comment by julietcohen** [comment]
Output:

```
Number of tiles with different number of duplicates: 235
Number of tiles with different number of rows: 0
```

Wow!

**Comment by julietcohen** [comment]
[dups_summary.csv](https://github.com/PermafrostDiscoveryGateway/viz-staging/files/11054890/dups_summary.csv)

Looks like neither Delta not Datateam consistently identified more/less duplicates than the other (there are both positive and negative non-zero values in the difference column)

**Comment by robyngit** [comment]
Did a quick little plot and it looks like indeed the differences are concentrated around where there is probably footprint overlap:

![dups_summary](https://user-images.githubusercontent.com/26600641/227365637-75cadf70-b91b-4d6d-a342-ca9a381f3f48.png)

```python
import pandas as pd
import geopandas as gpd
from shapely.geometry import box
from pdgstaging import TilePathManager
from matplotlib import pyplot as plt

path = "dups_summary.csv"
df = pd.read_csv(path)
tpm = TilePathManager(tms_id="WorldCRS84Quad")


def make_geom(row):
    tile = tpm.tile(row["x"], row["y"], row["z"])
    bb = tpm.get_bounding_box(tile)
    n, s, e, w = bb["top"], bb["bottom"], bb["right"], bb["left"]
    return box(w, s, e, n)


df["geometry"] = df.apply(make_geom, axis=1)

gdf = gpd.GeoDataFrame(df, geometry="geometry", crs=tpm.crs)

gdf.plot(column="num_dups_diff", legend=True)
plt.savefig("dups_summary.png")
plt.show()
```

**Comment by julietcohen** [comment]
Amazing! I was just wrangling the dataframe in R to figure out a trend in the duplicate flagging, plotting it right away was much more clever. Very informative

**Comment by julietcohen** [comment]
## Ideas for changes to `viz-staging`

Which files are identified as duplicates depends on order of files staged?
- Files staged serially: file 1 polygons → file 2 polygons → file 3 polygons (mostly)
- Files staged in parallel results in order of polygons identified as duplicates from all 3 input files much more jumbled

Solution:
- Execute `label_duplicates()` in `deduplicate_by_footprint()` after `clip_gdf()` (there is no point in labeling a polygon as a duplicate if we are going to clip it anyway), and execute these functions in the same order regardless if there is one or more input files that produced polygons in the geodatafame (basically remove [this code](https://github.com/PermafrostDiscoveryGateway/viz-staging/blob/331d52081c39598f887575e667cd17aaf71d09d1/pdgstaging/Deduplicator.py#L509-L518)?)
  - Alternative: label duplicates first then clip to footprint just before we save the tile, even if there is only 1 input file identified in the geodataframe. 
- execute `combine_and_deduplicate()` regardless if the tile is a new file or the polygons are being added to an existing file (code to change is [here](https://github.com/PermafrostDiscoveryGateway/viz-staging/blob/331d52081c39598f887575e667cd17aaf71d09d1/pdgstaging/TileStager.py#L405-L412), remove the line `if os.path.isfile(tile_path):`)
- Change predicate of clip to footprint (see options in [documentation](https://geopandas.org/en/stable/docs/reference/api/geopandas.sindex.SpatialIndex.valid_query_predicates.html))

**Comment by julietcohen** [comment]
Thank you, Robyn! I'm making these suggested changes in my branch.

**Comment by julietcohen** [comment]
## Testing branch on Datateam with 3 adjacent IWP files

Ran test of the new clipping and deduplication approach on Datateam. 

<details>
  <summary>workflow.py</summary>

```
# Use branch bug-17-clippingToFP of viz-staging.
# Main branch of viz-raster.
# Use env that pulls local versions of packages.
# env: arcade_layer

# Run staging, rasterization, and web-tiling on 3 adjacent
# files on Datateam and on Delta,
# and download the staged files from Delta to datateam.
# Then compare then number of tiles identified
# as duplicates in both staged directories using Robyn's script. 
# Also compare output web tiles.
# If the number of duplicates identified is the same, 
# and the web tiles look the same, the branch works as intended 
# for deduplication set to occur at raster step. 

from pathlib import Path

import pdgstaging
from pdgstaging import TileStager

import pdgraster
from pdgraster import RasterTiler

# logging
from datetime import datetime
import logging
import logging.handlers
import os

# removing files and dirs from past run(s)
import os
import shutil

# remove staged dir, log.log, and staging_summary.csv from 
# last test run if they exist
if os.path.exists("/home/jcohen/clip_fix_test/log.log"):
    os.remove("/home/jcohen/clip_fix_test/log.log")

if os.path.exists("/home/jcohen/clip_fix_test/staging_summary.csv"):
    os.remove("/home/jcohen/clip_fix_test/staging_summary.csv")

if os.path.exists("/home/jcohen/clip_fix_test/staged"):
    shutil.rmtree("/home/jcohen/clip_fix_test/staged")

# define 3 adjacent input shapefiles
base_dir = Path('/home/jcohen/iwp_russia_subset_clipToFP_PR/iwp')
filename = '*.shp'
# To define each .shp file within each subdir as a string representation with forward slashes, use as_posix()
# The ** represents that any subdir string can be present between the base_dir and the filename
input = [p.as_posix() for p in base_dir.glob('**/' + filename)]

# define their footprints
base_dir_fp = Path('/home/jcohen/iwp_russia_subset_clipToFP_PR/footprints')
fps = [p.as_posix() for p in base_dir_fp.glob('**/' + filename)]

# set up logging
handler = logging.handlers.WatchedFileHandler(
    os.environ.get("LOGFILE", "/home/jcohen/clip_fix_test/log.log"))
formatter = logging.Formatter(logging.BASIC_FORMAT)
handler.setFormatter(formatter)
root = logging.getLogger()
root.setLevel(os.environ.get("LOGLEVEL", "INFO"))
root.addHandler(handler)

# ------------- Staging ----------------

stager = TileStager({
  "deduplicate_clip_to_footprint": True, 
  "dir_input": "/home/jcohen/iwp_russia_subset_clipToFP_PR/iwp/", 
  "ext_input": ".shp",
  "ext_footprints": ".shp",
  "dir_footprints": "/home/jcohen/iwp_russia_subset_clipToFP_PR/footprints/", 
  "dir_staged": "staged/",
  "dir_geotiff": "geotiff/", 
  "dir_web_tiles": "web_tiles/", 
  "filename_staging_summary": "staging_summary.csv",
  "filename_rasterization_events": "raster_events.csv",
  "filename_rasters_summary": "raster_summary.csv",
  "filename_config": "config",
  "simplify_tolerance": 0.1,
  "tms_id": "WGS1984Quad",
  "z_range": [
    0,
    15
  ],
  "geometricError": 57,
  "z_coord": 0,
  "statistics": [
    {
      "name": "iwp_coverage",
      "weight_by": "area",
      "property": "area_per_pixel_area",
      "aggregation_method": "sum",
      "resampling_method": "average",
      "val_range": [
        0,
        1
      ],
      "palette": [
        "#66339952",
        "#ffcc00"
      ],
      "nodata_val": 0,
      "nodata_color": "#ffffff00"
    },
  ],
  "deduplicate_at": [
    "raster"
  ],
  "deduplicate_keep_rules": [
    [
      "Date",
      "larger"
    ]
  ],
  "deduplicate_method": "footprints"
})

for file in input:
    print(f"Staging {file}.")
    stager.stage(file)
# number of staged files should be 1056

print("Staging complete. Starting rasterization & web-tiling.")

# ------------- Rasterization ----------------

RasterTiler({
  "deduplicate_clip_to_footprint": True, 
  "dir_input": "/home/jcohen/iwp_russia_subset_clipToFP_PR/iwp/", 
  "ext_input": ".shp",
  "ext_footprints": ".shp",
  "dir_footprints": "/home/jcohen/iwp_russia_subset_clipToFP_PR/footprints/", 
  "dir_staged": "staged/",
  "dir_geotiff": "geotiff/", 
  "dir_web_tiles": "web_tiles/", 
  "filename_staging_summary": "staging_summary.csv",
  "filename_rasterization_events": "raster_events.csv",
  "filename_rasters_summary": "raster_summary.csv",
  "filename_config": "config",
  "simplify_tolerance": 0.1,
  "tms_id": "WGS1984Quad",
  "z_range": [
    0,
    15
  ],
  "geometricError": 57,
  "z_coord": 0,
  "statistics": [
    {
      "name": "iwp_coverage",
      "weight_by": "area",
      "property": "area_per_pixel_area",
      "aggregation_method": "sum",
      "resampling_method": "average",
      "val_range": [
        0,
        1
      ],
      "palette": [
        "#66339952",
        "#ffcc00"
      ],
      "nodata_val": 0,
      "nodata_color": "#ffffff00"
    },
  ],
  "deduplicate_at": [
    "raster"
  ],
  "deduplicate_keep_rules": [
    [
      "Date",
      "larger"
    ]
  ],
  "deduplicate_method": "footprints"
}).rasterize_all()

print("Rasterizing and web-tiling for all z-levels complete.")
```
</details>

Number of staged tiles : 3214
Number of geotiffs: 4376
Number of web tiles: 4376

<img width="758" alt="image" src="https://user-images.githubusercontent.com/88209596/233150027-cbf5397d-624b-4e66-9262-4ba00b6eb11d.png">

Deduplication where footprints overlap is working. Clipping to footprint was executed according to the log, but seems like those polygons were not removed.

**Comment by julietcohen** [comment]
Removing the above code chunk did not resolve the issue that the polygons that were flagged as duplicates because they fell outside the footprint were still not removed. So there must be multiple places in the deduplication approach that overwrite those labeled polygons in the 'duplicated' column of the tile with _just_ the duplicates that were identified from overlapping footprint areas. I adjusted `dedupicate_by_footprint()` so that the `to_remove` list that is created at the start is immediately populated with the subset of the input gdf that represents the polygons that were labeled as duplicates earlier, when we executed `clip_gdf` and then `label_duplicates` the first time. 

```python
    gdf = gdf.copy()

    # `to_remove` list will hold the polygons that fit either of the criteria:
    # 1. were previously labeled True for `duplicated` col because poly
    # fell outside footprint
    # 2. are labeled as True for `duplicated` col within this function
    # based on overlap of footprints
    to_remove = []

    # First, add the polygons that were already labeled as duplicates because
    # fell outside footprint
    known_dups = gdf[gdf['duplicated'] == True]
    to_remove.append(known_dups)

    logger.info(f"After initially adding the previously identified dups df to list to_remove, length is {len(to_remove)}\nand some values in list are {to_remove[0:10]}.")

    # Will hold the polygons that defined the footprint intersections
    intersections = []
```

The output webtiles looked the same. The search continues!

**Comment by julietcohen** [comment]
## Deduplication successful for both polygons that fall outside footprint and those in overlapping footprint regions

<img width="842" alt="image" src="https://user-images.githubusercontent.com/88209596/234066309-3d2c0ede-31e5-4cfb-af2a-899e97a57f3e.png">

**Comment by julietcohen** [comment]
## Changing predicate for clipping to footprint

After successfully clipping to footprint and removing duplicates where the footprints overlap, the next step is to change the geopandas' sjon predicate for clipping to footprint in order to eliminate the light strips we see between some adjacent shapefiles when you zoom in and look real closely:

<img width="1644" alt="within" src="https://user-images.githubusercontent.com/88209596/234716259-783695d0-a0fd-44c3-84b2-a281bf6a7dec.png">

Instead of using 'within', which only retains the polygons that fall completely inside the footprint boundary, we want to retain polygons that fall completely within the footprint boundary and those that partially fall within the boundary, but hang over the edge. Changing it to 'intersects' did the trick!
<img width="1713" alt="image" src="https://user-images.githubusercontent.com/88209596/234715920-0ef94e9c-4d0e-4cdc-a800-19b3b910a912.png">

**Comment by robyngit** [comment]
Wow, that looks fantastic @julietcohen !!!

**Comment by julietcohen** [comment]
Thanks Robyn!

**Comment by julietcohen** [comment]
## Viz-workflow on Delta (with 1 CPU node) executes deduplication (almost?) as expected

<img width="897" alt="image" src="https://user-images.githubusercontent.com/88209596/235797736-213943de-3c77-434e-ade8-b2de7b911e8f.png">

Note: it is noticeable that there is a light strip (that we noticed we _weren't_ seeing anymore after switching the `clip_gdf` predicate from 'within' to 'intersects'), along the border of the bottom 2 tiles. I wonder why the border of the upper 2 tiles does not show this, while the bottom does. 
<img width="1861" alt="dedup_line" src="https://user-images.githubusercontent.com/88209596/235798233-5330aa37-4880-49af-8fd8-db7ed0a35f56.png">


Regardless, this is a big improvement in the deduplication using the Ray workflow.

**Comment by julietcohen** [comment]
Next run on Delta:

- increase number of adjacent input files because we want to test running the workflow in parallel on multiple nodes. With such little input data as shown above (3 shapefiles and their footprints), the viz-workflow only runs on 1 node during staging, and most of rasterization is on 1 node too, because by the time one or multiple worker nodes fire up, the head node has already rasterized most of the files
- insert print statement (because `logging.info()` is still largely not working) to state when clipping is happening and the predicate used

**Comment by julietcohen** [comment]
## Parallel ray workflow run on Delta: all Wrangel Island IWP detections

**Setup:**
- 4 CPU nodes
- 104 shapefiles (all of `high/russia/226_227_iwp` dir)
- changed palette from 2 colors (yellow & pink) to 9 hex colors ranging from gray-red to bright red:
```
       ["#a15e5e", # gray-red
        "#aa5555",
        "#b44b4b",
        "#be4141",
        "#c93636",
        "#d42b2b",
        "#e31c1c",
        "#f50a0a",
        "#ff0000" # bright red]
```
- `viz-staging` branch `bug-17-clippingToFP`
- `viz-raster` branch `bug-ID-notRecognized`

**staging:**
- during staging, only 2 nodes were producing staged files in their respective `/tmp` directories (head, and 1 worker)
- 13 min
- head node was at ~20% CPU, ~22% mem
- worker node that staged files was at ~8% CPU, ~17% mem
- idle worker nodes we at ~0.2% CPU, ~5% mem
- after merging, 50,193 staged files

**raster highest:**
- during rasterization highest, all 4 nodes were producing geotiff files in their respective `/tmp` directories
- head node was at ~25% CPU, ~16% mem
- worker nodes were at ~25% CPU, ~15% mem but producing geotiffs slower than head node
- interesting errors written, which don't make sense to me because the geotiff dir is created initially in each nodes' `/tmp` and is empty, but seems like a couple geotiffs are attempted to be written multiple times?
<img width="1309" alt="Pasted Graphic 3" src="https://user-images.githubusercontent.com/88209596/236019157-bafdb3fe-6b3f-47b7-8f6e-5da5e568d3e3.png">

- total number of geotiffs highest produced: 50,184 (so like 9 errored/did not get written)
- 6 min

**raster lower:**
- 68,517 geotiffs produced for all z-levels
- 2 minutes
- did have to remove a few incorrectly formatted lines from the top of `raster_summary.csv` before executing web tiling, as usual

**web tiles:**
- 68,517 web tiles produced for all z-levels
- 6 min

Final output:
<img width="1509" alt="image" src="https://user-images.githubusercontent.com/88209596/236048047-c4fb7552-890f-4664-8e23-286060a353ef.png">
<img width="1861" alt="image" src="https://user-images.githubusercontent.com/88209596/236048405-5c8772c5-79e4-4e9c-a861-4808263eac03.png">
<img width="1858" alt="image" src="https://user-images.githubusercontent.com/88209596/236048450-a9f7ccaa-279a-4f68-b7c4-7534487bb3e5.png">

**Comment by julietcohen** [comment]
Relics of edge effects are still visible between tiles but only in the form of clipping to footprint, not where footprints overlap. However, these edge effects are not present between every tile, just every once in a while you spot one when scrolling around. 
<img width="1836" alt="edge" src="https://user-images.githubusercontent.com/88209596/236053321-b5cd9b19-acad-4d16-a061-067cf83d82a6.png">

Then when you zoom in on some of the border strips, it's clear there are still polygons that were indeed retained that just fell on the edge of the footprint, which is what we want (we just want to remove the polys that fall _completely_ outside the footprint). I wonder if these empty strips that are present between some tiles are unavoidable, no matter the predicate used.

<img width="1848" alt="edge2" src="https://user-images.githubusercontent.com/88209596/236053348-42e4c15a-d903-44bb-825c-84eda1882ac6.png">

**Comment by julietcohen** [comment]
Additional note: `raster_summary.csv` still shows the maximum value for `iwp_coverage` statistic as >1. I still have to remove 3 or 4 incorrectly formatted rows and re-upload the csv in between raster highest and web tiling. 
<img width="1024" alt="max_above_1" src="https://user-images.githubusercontent.com/88209596/236271963-4ffb166e-f469-4e2c-b2cc-f59f69868768.png">

The values >1 occurs on datateam (without parallelization) as well.

**Comment by julietcohen** [comment]
## Comparing colors and transparency for IWP layer

### green, 0.10 - full color
<img width="1405" alt="image" src="https://user-images.githubusercontent.com/88209596/236331661-3a830009-953a-4ec8-be77-42ac7f2b91be.png">

<img width="1766" alt="image" src="https://user-images.githubusercontent.com/88209596/236331795-a9b76567-d72d-42bd-88c3-c235b7627ac6.png">

<img width="1762" alt="image" src="https://user-images.githubusercontent.com/88209596/236331839-43373e4c-38fc-40ac-9d0f-9811cc6be568.png">

**Comment by julietcohen** [comment]
## purple, 0.10 - full color

<img width="1427" alt="image" src="https://user-images.githubusercontent.com/88209596/236585375-b9e32334-c09e-4fd1-a916-9732a0741101.png">
<img width="917" alt="image" src="https://user-images.githubusercontent.com/88209596/236585415-123d74b3-c453-4682-b82a-ad513396c2b5.png">
<img width="1852" alt="image" src="https://user-images.githubusercontent.com/88209596/236585427-4e75e8e9-6826-4778-9b17-2b9d55ec3510.png">

**Comment by julietcohen** [comment]
## red, 0.1 - full color 

<img width="1219" alt="image" src="https://user-images.githubusercontent.com/88209596/236586214-f888c7bb-8da4-499a-bf78-5ac4d7646d2c.png">
<img width="1721" alt="image" src="https://user-images.githubusercontent.com/88209596/236586242-e773ebe4-3b86-4aeb-9031-b56e404483f0.png">
<img width="1757" alt="image" src="https://user-images.githubusercontent.com/88209596/236586253-42b4db71-9487-44ef-999d-e570da39be5e.png">

**Comment by julietcohen** [comment]
## yellow, 0.1 - full color

<img width="1313" alt="image" src="https://user-images.githubusercontent.com/88209596/236586462-0defa46c-c9f2-4f2d-a626-85e94b983ffe.png">
<img width="1769" alt="image" src="https://user-images.githubusercontent.com/88209596/236586488-04e7216d-e1ba-4fae-b7bf-1c38650717ea.png">
<img width="1750" alt="image" src="https://user-images.githubusercontent.com/88209596/236586499-f56f3ada-385a-453e-9ffd-d34cba1be894.png">

**Comment by julietcohen** [comment]
Plotting the footprints for these input files shows the area that was covered by input files (with partial transparency to show where footprints overlap). This shows that the strip of "missing" IWP on the map is a result of the lack of input data for that area, not errors in the visualization workflow.
<img width="684" alt="fp" src="https://user-images.githubusercontent.com/88209596/236952109-01aab208-067a-4d3b-a5bc-4c2fa1b63f45.png">

**Comment by mbjones** [comment]
Nice -- let's talk to Elias and Chandi and figure out if this missing data is unexpected from their perspective.

**Comment by julietcohen** [comment]
The strip of missing tiles shown above is due to the [180th meridian](https://en.wikipedia.org/wiki/180th_meridian), which falls on Wrangle Island. See here:
![image](https://github.com/PermafrostDiscoveryGateway/viz-staging/assets/88209596/baa9e447-e466-4eb9-94d8-450835a93cbb)

**Comment by julietcohen** [comment]
I finalized the changes to the new `viz-staging` deduplication approach today. This new approach runs smoothly on both Delta as well as Datateam with identical output. The logging in both the ray workflow and simple `viz-workflow` now works as well!

The branch `bug-17-clippingToFP` has also been tested to confirm it works when the `dedup_method` is set to `None` and `clipping_to_footprint` is set to `False` in the config (the result is what we expect with that config):

<img width="542" alt="image" src="https://github.com/PermafrostDiscoveryGateway/viz-staging/assets/88209596/d23c54ea-6246-42c5-bf60-5d4e0e3efd92">


The `neighbors` deduplication approach has been modified slightly as well (essentially removing the `label_duplicates()` as a separate function, and the necessary parts have been tailored to and integrated into the `neighbors` method), but the `neighbors` method has not been thoroughly tested yet. We decided that it is higher priority to process all the IWP data before we test it. As a result, the the `viz-staging` package with these changes has been documented that the `neighbors` approach is not to be used until a release after 0.1.0. The release 0.1.0 is tailored to the `footprints` dedup method.
